### PR TITLE
INF-173: action vue build to github actions on pull request

### PR DESCRIPTION
**Pull request by ctucker3**
Adds checks to bi-web on pull request. A bunch of our tests are broken, so the tests aren't added to this yet. But, those should be easy to add once they are fixed. 

Test runs a `vue-cli-service build` and a `npm audit` check. The `vue-cli-service` runs a linter and type check on the project and fails if there are any errors there. You will see some fixes to the types in existing files because I couldn't get the build to run without fixing those. But, these changes are a good example of the things that will be caught in the future!

Also includes changes to the `package-lock.json` file to remove the bulk of the vulnerabilities. 

I have add a merge condition that the tests must pass for bi-web in order to be merged. 

I have add a commit to the PR for both failure scenarios:

1. Vulnerable libraries
2. Vue build error

You can see the results of both tests in the actions:

https://github.com/Breeding-Insight/bi-web/actions

Or, if you look down below in the list of commits it will shows X's and checkmarks for successes and failures on each commit to the existing PR.

## Design 
I made a custom `npm audit` check because `npm audit` returns an exit code of 1 if there are any vulnerabilities. There are some promising looking flags like `--only=prod` that don't work for `npm audit`, only `npm audit fix`. There is also `--audit-level=<serverity>` that works, but treats dev and prod errors the same. This seemed like it would cause unneeded burden on our dev libraries since the security on those shouldn't need to meet the same standard as our prod libraries. 

The custom error checking accepts parameters for the maximum acceptable severity error level for dev and for prod separately. It's nothing too complicated, it just parses the return and checks the errors. I currently have dev set to high because our dev libraries shouldn't be in our production build, and prod set to moderate, because the meaning of moderate severity is: `Address as time allows`. I'm open to discussion on this.

### CODE REVIEW

**Comment by timparsons** [comment]
[INF-173 - Configure GitHub Actions to automatically build bi-web](https://trello.com/c/GxUVc708/260-inf-173-configure-github-actions-to-automatically-build-bi-web)

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v1
+        with:
+          node-version: ${{ matrix.node-version }}
+      - name: Install dependencies
+        run: npm install
```
Comment (dmeidlin) [inline .github/workflows/build.yml:41]: "npm ci" should be used instead since it builds off of package-lock.json; whereas, "install" uses package.json.

**Comment by timparsons** [comment]
> Not a reviewer on this but was looking at the github actions to see something and noticed there are a bunch of changed code files in here. Was that intended/needed for the build and will stuff have to be changed since some of that code is old?

@nickpalladino Only commenting because Chris is off today, and we just happened to talk about this yesterday.  The code changes were needed to get the linter checks to pass.

File: .github/workflows/build.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ctucker3) [reply .github/workflows/build.yml]: Changed and pushed

**Comment by ctucker3** [comment]
> The build.js script looks good for setting risk thresholds for prod and dev. We are scheduled next Friday to discuss how to handle these audit results, and right now I am not convinced there is a need to filter audit results based on level for either dev or production.
> 
> Any audit flags for dev can be safely ignored because the vulnerable packages won't be bundled into the production build, and you don't want any known vulnerabilities in production regardless of severity level, so there is no need to filter by severity level for production packages.
> 
> The response from the npm team on a github issue requesting the --audit-level option was that the audit levels are meant only to suggest priority:
> from npm dev in repo issue: [npm/npm#20596](https://github.com/npm/npm/issues/20596)
> "Issues of any severity should really be fixed, the severity of a vulnerability assessment should only serve to prioritize the fixes."
> 
> If we want a passing test to be no security issues in prod, ignoring dev, then the following npm command can help:
> `bi-web_$npm audit fix --dry-run --json --only=prod `
> `{ "added": [], "removed": [], "updated": [], "moved": [], "failed": [], "warnings": [ "The package bulma is included as both a dev and production dependency." ], "funding": "53 packages are looking for funding", "elapsed": 5306 }`
> 
> A passing test results from parsing the json for empty arrays in the first six keys. This also checks for warnings that we may want to address. In this case bulma should be removed as a prod dependency.
> In other cases we may want to ignore the warning, so having an option to ignore warnings seems good to have.

Thanks for checking it out! I would propose that we create a card to explore the option you list above. I added a `none` option for the vulnerability check in the case that we decide that there should be no vulnerabilities for prod. The severity check as it is works for what we need it for, and I don't want to get stuck on this card too much longer. Are you good with that?

It is outside the scope of this card, but I think bulma should be removed as a dev dependency, not a prod dependency. I believe we are going to need that for production.

**Comment by dmeidlin** [comment]
Sounds good.

**Pull request merged**